File: macos/VPNRouter/VPNRouter/Shared/Rules/DomainRuleExpander.swift
import Foundation

enum DomainRuleExpander {
    private nonisolated static let relatedDomains: [String: [String]] = [
        "youtube.com": [
            "googlevideo.com",
            "ytimg.com",
            "youtubei.googleapis.com",
            "ggpht.com",
            "youtube-nocookie.com"
        ],
        "netflix.com": [
            "nflxvideo.net",
            "nflximg.net",
            "nflximg.com",
            "nflxso.net",
            "nflxext.com"
        ]
    ]

    nonisolated static func expand(_ rules: [DomainRule]) -> [DomainRule] {
        var expanded = rules
        var domains = Set(rules.map { normalize($0.domain) })

        for rule in rules where rule.enabled {
            let domain = normalize(rule.domain)

            for (root, relatedDomains) in Self.relatedDomains where matchesRoot(domain, root: root) {
                for relatedDomain in relatedDomains where domains.insert(relatedDomain).inserted {
                    expanded.append(DomainRule(
                        profileId: rule.profileId,
                        domain: relatedDomain,
                        includeSubdomains: true,
                        enabled: true,
                        createdAt: rule.createdAt
                    ))
                }
            }
        }

        return expanded
    }

    nonisolated static func normalize(_ domain: String) -> String {
        domain
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "."))
            .lowercased()
    }

    private nonisolated static func matchesRoot(_ domain: String, root: String) -> Bool {
        domain == root || domain.hasSuffix(".\(root)")
    }
}

File: windows/VpnRouter.Core/Rules/DomainRuleExpander.cs
namespace VpnRouter.Core.Rules;

public static class DomainRuleExpander
{
    private static readonly IReadOnlyDictionary<string, string[]> RelatedDomains =
        new Dictionary<string, string[]>(StringComparer.OrdinalIgnoreCase)
        {
            ["youtube.com"] =
            [
                "www.youtube.com",
                "googlevideo.com",
                "ytimg.com",
                "youtubei.googleapis.com",
                "ggpht.com",
                "youtube-nocookie.com"
            ],
            ["netflix.com"] =
            [
                "www.netflix.com",
                "nflxvideo.net",
                "nflximg.net",
                "nflximg.com",
                "nflxso.net",
                "nflxext.com"
            ]
        };

    public static IReadOnlyList<DomainRule> Expand(IReadOnlyList<DomainRule> rules)
    {
        var expanded = new List<DomainRule>(rules);
        var domains = rules
            .Select(rule => rule.Domain)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

        foreach (var rule in rules.Where(rule => rule.Enabled))
        {
            foreach (var mapping in RelatedDomains)
            {
                if (!MatchesRoot(rule.Domain, mapping.Key))
                {
                    continue;
                }

                foreach (var relatedDomain in mapping.Value)
                {
                    if (domains.Add(relatedDomain))
                    {
                        expanded.Add(new DomainRule(
                            Guid.NewGuid(),
                            rule.ProfileId,
                            relatedDomain,
                            IncludeSubdomains: true,
                            Enabled: true,
                            CreatedAt: rule.CreatedAt));
                    }
                }
            }
        }

        return expanded;
    }

    private static bool MatchesRoot(string domain, string root) =>
        string.Equals(domain, root, StringComparison.OrdinalIgnoreCase)
        || domain.EndsWith("." + root, StringComparison.OrdinalIgnoreCase);
}
